dashboard/app/models/user.rb

```diff
@@ -615,7 +615,7 @@ def email_required?
   end
 
   def username_required?
-    provider == User::PROVIDER_MANUAL
+    provider == User::PROVIDER_MANUAL || username_changed?
```

Comment:
AFAIK, there isn't any reason not to require a username for all accounts. As long as they aren't going away (as much as I'd like them to, they definitely aren't), we might as well require them. Note that my code for account purging (https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/models/user.rb#L1489) anonymizes the username rather than clearing it entirely.

That said, we should confirm with Product that it makes sense to do so. Am happy to own making this invariant happen in all environments.